sources/models/src/modeled_types/kubernetes.rs

```diff
@@ -1324,3 +1307,98 @@ pub struct CredentialProvider {
     image_patterns: Vec<SingleLineString>,
     cache_duration: Option<KubernetesDurationValue>,
 }
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// KubernetesMemoryReservationKey represents a string that contains a valid Kubernetes memory
+/// resource reservation key.
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KubernetesMemoryReservationKey {
+    inner: String,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum MemoryReservationResources {
+    #[serde(rename = "memory")]
+    Memory,
+    #[serde(rename = "hugepages-2Mi")]
+    HugePages2Mi,
+    #[serde(rename = "hugepages-1Gi")]
+    HugePages1Gi,
```

Comment:
Can you confirm that `hugepages-2Mi` and `hugepages-1Gi` are not mutually exclusive? I was originally under the impression that they might be, but from the docs I've drawn the impression that these are treated as three distinct types of memory resource limits.

Reply:
The docs aren't great here, but these are treated as distinct types of memory resource limits. The user would need to determine what is available by using `kubectl describe node` to see what is reported as the available values:

```
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                    Requests    Limits
  --------                    --------    ------
  cpu                         325m (16%)  0 (0%)
  memory                      140Mi (2%)  340Mi (5%)
  ephemeral-storage           0 (0%)      0 (0%)
  hugepages-1Gi               0 (0%)      0 (0%)
  hugepages-2Mi               0 (0%)      0 (0%)
```

Reply:
@foersleo or @markusboehme - can you tell if anything else is needed to [support using hugepages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt)?

From what I've read, this would require passing in kernel parameters on boot, but I'm not clear on the full implications of that and how things are handled with Bottlerocket.

Reply:
You're right, some kernel configuration is required. At some point the kernel needs to be told how many huge pages of which size to set aside. This does not necessarily have to happen at boot, it can also be done at any later point in time via procfs or sysfs. However, a huge page is a long stretch of physically contiguous memory, and physical memory will get fragmented over the course of normal system operation. This means the chance of successfully allocating huge pages decreases over time. Hence, it's best to do so early.

In Bottlerocket, the above translates to setting the kernel command line via the user data mechanism or setting 
corresponding values in sysfs. For example, to set aside 1024x 2 MiB huge pages at boot, one might use

```toml
[settings.boot]
reboot-to-reconcile = true

[settings.boot.kernel]
hugepages = ["1024"]
```

To not pay the price of a reboot (modifying the kernel command line happens after first boot), an alternative would be to set a sysctl (this happens later in the boot, though, after Bottlerocket's corndog is running):

```toml
[settings.kernel.sysctl]
vm.nr_hugepages = 1024
```

Or with a bootstrap container writing `/sys/kernel/mm/hugepages/hugepages-*/nr_hugepages`, here with control over which size huge pages to set aside.

Some implications/things to watch out for:

* Huge pages set aside this way cannot be used for other, regular allocations. Corollaries: 1. If no pods use them, they're dead memory. 2. They add to memory pressure. Workloads can run out of memory earlier if they're not appropriately sized ("requests" in pod spec).
* Huge pages cannot be swapped out. Not a problem for Kubernetes, since swap is not supported anyway.
* Requesting too many huge pages at boot time (i.e. close to or more than the system memory) prevents the system from booting.